File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.java
class Solution {
    public int singleNonDuplicate(int[] nums) {
        //Binary Search
           // -> Boundarty checks
           //  -> Pair index property
           //     -> unique element property
           //      -> partition property
        int n = nums.length;
        int low = 0, high = n-1,mid = 0;
        if(high == 0) return nums[0];
        //Boundart Checks
        else if(nums[0] != nums[1]) return nums[0];
        else if(nums[high] != nums[high - 1])  return nums[high];
        
        while(low <= high) {
            mid = (low + high) / 2;
            // unique element check
            if(nums[mid] != nums[mid - 1] && nums[mid] != nums[mid +1]) {
                return nums[mid];
            }
            // Pair index Property
            if(mid % 2 != 0 && nums[mid] == nums[mid - 1] ||  mid % 2 == 0 && nums[mid] == nums[mid + 1]){
                low = mid + 1; // the order of starting at even and ending at odd is not disrupted
            }else {
                high = mid - 1;
            }
        }
        return -1;
    }
}

File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.cpp
class Solution {
public:
    int singleNonDuplicate(vector<int>& num) {
        int n = num.size();
        if(n == 1) return num[0];
        for(int i = 0; i < n; i++) {
            if(i == 0) {
                if(num[i] != num[i+1]) return num[0];
            } else if( i == n) {
                if(num[i] != num[i-1]) return num[i];
            } else {
                if(num[i] != num[i+1] && num[i] != num[i-1])  {
                    return num[i];
                }
            }
        }
        return -1;
    }
};